drivers/platform/maxim/maxim_spi.c

```diff
@@ -0,0 +1,186 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/************************* Include Files **************************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include <errno.h>
+#include "spi.h"
+#include "spi_extra.h"
+#include "no-os/spi.h"
+#include "no-os/util.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Initialize the SPI communication peripheral.
+ * @param desc - The SPI descriptor.
+ * @param param - The structure that contains the SPI parameters.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t max_spi_init(struct spi_desc **desc, const struct spi_init_param *param)
+{
+	if (!param)
+		return -EINVAL;
+
+	int32_t ret;
+	struct spi_desc *descriptor = calloc(1, sizeof(*descriptor));
+
+	if (!descriptor)
+		return -ENOMEM;
+
+	descriptor->device_id = param->device_id;
+	descriptor->max_speed_hz = param->max_speed_hz;
+	descriptor->chip_select = param->chip_select;
+	descriptor->mode = param->mode;
+	descriptor->bit_order = param->bit_order;
+	descriptor->platform_ops = &max_spi_ops;
+
+	if (descriptor->device_id == 0)
+		ret = SPI_Init(SPI0A, descriptor->mode, param->max_speed_hz);
+	else
+		ret = SPI_Init(SPI1A, descriptor->mode, param->max_speed_hz);
+
+	if (ret != 0)
+		goto error;
+
+	*desc = descriptor;
+
+	return 0;
+err:
+	free(descriptor);
+
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by spi_init().
+ * @param desc - The SPI descriptor.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t max_spi_remove(struct spi_desc *desc)
+{
+	if (!desc)
+		return -EINVAL;
+	free(desc);
+
+	return 0;
+}
+
+/**
+ * @brief Write and read data to/from SPI.
+ * @param desc - The SPI descriptor.
+ * @param data - The buffer with the transmitted/received data.
+ * @param bytes_number - Number of bytes to write/read.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t max_spi_write_and_read(struct spi_desc *desc,
+			       uint8_t *data,
+			       uint16_t bytes_number)
+{
+	if (!desc || !data)
+		return -EINVAL;
+
+	uint8_t *tx_buffer = data;
+	uint8_t *rx_buffer = data;
+	spi_req_t req;
+
+	req.bits = 8;
+	req.ssel = 0;
+	req.tx_data = tx_buffer;
+	req.tx_num = 0;
+	req.rx_data = rx_buffer;
+	req.rx_num = 0;
+	req.deass = 0;
+	req.width = SPI17Y_WIDTH_1;
+	req.len = bytes_number;
+
+	if (desc->device_id == 0)
+		SPI_MasterTrans(SPI0A, &req);
```

Comment:
check error codes for all HAL functions in spi.c and translate them into no-os error codes